### ISSUE TITLE
Brackets improperly matched with inline statements in Ruby

### ISSUE THREAD

**Issue by x9prototype** [facelessuser/BracketHighlighter#452]
## Description

Bracket Highlighter doesn't ignore inline statements in Ruby.  For example, having the following function would mis-match the "end" to the inline if-statement and not the function definition.

The following code will properly highlight def and end
```ruby
def isSSL(url)
  uri = URI.parse(url)
end
```
![correct_parse](https://user-images.githubusercontent.com/9808370/32815727-4e2026ee-c96a-11e7-90ba-e40f39aec6b3.png)

Adding an inline if statement breaks proper end statement matching
```ruby
def isSSL(url)
  uri = URI.parse(url)
  return if uri.scheme != 'https'
end
```
![incorrect_parse1](https://user-images.githubusercontent.com/9808370/32815726-4e07022c-c96a-11e7-8521-b8737813d4c5.png)
![incorrect_parse2](https://user-images.githubusercontent.com/9808370/32815725-4df1bb88-c96a-11e7-919b-55e89d6b7940.png)

## Support Info

- ST ver.: 3143
- Platform: osx
- Arch: x64
- Plugin ver.: 2.26.0
- Install via PC: True
- mdpopups ver.: 3.1.3
- backrefs ver.: 2.1
- markdown ver.: 2.6.9
- pygments ver.: 2.1a0
- jinja2 ver.: 2.8

## Steps to Reproduce Issue

1. Copy the above code from the description into Sublime.
2. Save the file as whatever.rb (or make sure to enable ruby syntax highlighting
3. Click on the `if` or the `end` statement and notice that the `end` no longer highlights with the `def`

**Comment by facelessuser** [comment]
Is this even valid? Keep in mind I'm not a ruby programer, but why would you ever code like that? Wouldn't it make more sense to do something like this:

```ruby
def isSSL(url)
  uri = URI.parse(url)
  return uri.scheme != 'https'
end
```

Wouldn't `uri.scheme != 'https'` yield a boolean?

I would expect either of these two for valid ruby if statements:

```ruby
if
# code
end
```

Or maybe

```ruby
condition ? if_true : if_false
```

But even if Ruby allows a single line `return if uri.scheme != 'https'`, it seems unnecessary (to me, a non Ruby programmer).  And how do I even begin to to tell when an `if condition` is a one line condition or when it is not? BH uses simple patterns to find an opening or closing bracket.  It does not take in consideration via context how it is nested in parent blocks.

I might think that if `if` followed a `return`, maybe it is a one liner, but based on a recent issue filed here: https://github.com/facelessuser/BracketHighlighter/issues/425, I also know that Ruby allows even more wackiness:

```ruby
def isSSL(url)
  uri = URI.parse(url)
  return if uri.scheme != 'https'
    true
  else
    false
  end
end
```

In short, BH has no way of looking at an `if` in this scenario and telling that it is expected to be evaluated without an `end`.  But I would argue further, IMHO, that there are better ways to represent what you mean to in your code regardless of what Ruby allows. Just use `return uri.scheme != 'https'`.  It makes much more sense (at least to me), but maybe I'm missing something?

Anyways, I don't use Ruby, so I would not be implementing a fix for this and require a pull request from the community as Ruby users understand Ruby better than me.  But based on what I see here, I'd argue that there is no way BH could tell this was meant to be a one line `if` as it simply doesn't have access to the context necessary to determine this.

**Comment by x9prototype** [comment]
So, the example was simplified to illustrate the problem, but a common design pattern to check data is `return if` or `return unless` in ruby (unless is basically a !if and will match with `end` "brackets" as well).  See the [Ruby Style Guide](https://github.com/bbatsov/ruby-style-guide#no-nested-conditionals)'s No Nested Conditionals Section ([rubocop](https://github.com/bbatsov/rubocop) a ruby linter follows this style guide).  I have pulled some code from the style guide and pasted it below:

```ruby
# bad
def compute_thing(thing)
  if thing[:foo]
    update_with_bar(thing[:foo])
    if thing[:foo][:bar]
      partial_compute(thing)
    else
      re_compute(thing)
    end
  end
end

# good
def compute_thing(thing)
  return unless thing[:foo]
  update_with_bar(thing[:foo])
  return re_compute(thing) unless thing[:foo][:bar]
  partial_compute(thing)
end

# bad
[0, 1, 2, 3].each do |item|
  if item > 1
    puts item
  end
end

# good
[0, 1, 2, 3].each do |item|
  next unless item > 1
  puts item
end
```

**Comment by facelessuser** [comment]
> So, the example was simplified to illustrate the problem

Fair enough

> Is there a way to have it match and ignore a return if or return unless?

Yes we can ignore `return if` as it used to be ignored before #425. A matter of fact, you'll find `unless` is still ignored in the scenarios you posted above.  Basically, I would just revert that change. I was actually horrified by the example in #425, but I figured it was a Ruby thing.

People unfortunately don't usually care to put in the time to help offer a good solution or help me come to a good understanding of the issue. The guy in #425 took a stab at it then went dark without explaining why. So I ended up just throwing in a solution without understanding the rippling effects. I don't have the time to learn the nuances of languages I don't use.  This is why I rely on the community for such languages.

If it is *more* proper to ignore an `if` following a `return`, then we can do that.  I just need to know what's best. I'm limited to how much context I can gather with regex, but ignoring a preceding return is no big deal.

> And how does it internally handle else, could it match against else end by itself with no if or unless? That may also solve your example from issue #425

It doesn't handle else.  BH *only* looks for opening brackets and closing brackets in a sliding window of your view.  It then matches them up until it is confident that the brackets are the closest ones around your cursor.  `else` is neither an opening or closing.  Keep in mind that these conditional caps are just piggy backing off the bracket logic as they are not *really* brackets.

In general, there is no deep analysis of a language.  It uses syntax scope and regular expression.  You can get some additional logic if you implement a `bh_plugin`, so tags is a special exception as I built a fairly elaborate plugin for it.

For the most part tags are well defined.  You have tags that must have closings.  You have tags that can be self closing.  You have tags that *must* be self closing.  And then you have tags that *can* have closings omitted in certain cases -- this is where BH reaches it's limits.  BH is more concerned with getting the job done as soon as it can so not to bog down the editor. It has been coded with no knowledge of parents, so in the tag example (where a tag is allowed to have a closing omitted in special cases), if a closing cannot be found for it, BH just assumes it is okay to omit the tag.

Tags is the most elaborate case, all others bracket logic is pretty simple with occasional hooks with plugins to do mundane stuff.

**Comment by x9prototype** [comment]
An interesting thing about Ruby is that any statement/expression returns the value of the last evaluated expression.  This could lead to code that looks like this:
```ruby
def numberFour
  4
end
```
Due to that, it is common to have a `return case` `end` pairing (or return any result of a block execution), and it would be valid (since the result of the case block would be what you want to return).

It is less common however to see a `return if` `end` or `return unless` `end`, so I would lobby for reverting the change on `return if` `return unless`, but if possible leaving it for other generic cases of `return <something_else>`.

There is a case of `return x if x > 5` case, that should be handled the same as `return if whatever`.

**Comment by facelessuser** [comment]
> There is a case of return x if x > 5, that should be handled the same as return if whatever.

This would be ignored after I revert the recent change.  Originally, if `if` was preceded by anything that was not whitespace, it was assumed to have no `end`.

So basically I'm looking for something more clear cut.  Which of these should be allowed to follow a return and still have an end? I've crossed out the ones I think should not.

- ~~if~~
- begin
- case
- ~~private~~
- ~~public~~
- ~~protected~~
- for
- until
- ~~unless~~
- while
- class
- module
- ~~do~~

**Comment by x9prototype** [comment]
The following usually don't come with an `end` if using `return`
`if`, `unless`

The following should expect an `end` statement
`do`, `class`, `begin`, `case`, `loop`, `for`, `until`, `while`, `module`

**Comment by facelessuser** [comment]
> The following should expect an end statement
do, class, begin, case, loop, for, until, while, module

1. All of these can follow a return? If so, then when I get some time I can throw something together to cover this case.  I'll post it here as I'll need you to test it to verify it.

2. From what little I understand, `loop` is usually associated with `do`?  If so, in that case `do` would get paired with the `end` and not `loop` and that is already covered.

**Comment by x9prototype** [comment]
`do` is usually associated with an enumerator (`loop`, `each`, etc...) and wouldn't come by itself, but is never without an `end`, and that makes sense.

**Comment by facelessuser** [comment]
Okay, so in the end it looks like we are just allowing: `class`, `begin`, `case`, `for`, `until`, `while`, and `module` to follow return with an `end`.  `do` doesn't ever directly follow return and is already handled.

I don't know when I'll get to this, but when I do, I'll post a solution here to ensure it is a working solution.  Thanks for taking time to clearly outline this.

**Comment by facelessuser** [comment]
This is issue is pending review of #458.  Please review and test changes to help get this in the next release.

**Comment by facelessuser** [comment]
Meh, #458 meets the specs we discussed, and it the cases described in #425 still works, and the issue you complained about works.  I'm going to go ahead and push the changes just so I can be done for the day with BH work.  It won't regress anything we didn't already talk about.  I figure any issues that it remain after it were already there before the pull or we explicitly deemed here as acceptable (like rejecting if following a return)

Thanks again for taking the time to talk this out so we could get this as good as possible. If you stumble on any other issues, we can address them in separate issues.
